fix: pass params when CRM contact owner fetch fails

joeauyeung opened the pull request:
## What does this PR do?

This PR fixes an issue where CRM contact owner fetch failures in routing forms would cause the entire form submission to fail, preventing other important params from being passed to the booking URL.

**Problem**: When `getRoutedUrl` calls `handleResponse`, if the CRM contact owner fetching fails (via `routerGetCrmContactOwnerEmail`), the entire function throws an error. This means params like `formResponseId` and `teamMembersMatchingAttributeLogic` are lost, causing unnecessary refetching when users visit the event URL.

**Solution**: Wrap the CRM contact owner fetching in a try-catch block. On error, log it and set CRM values to null, allowing the form submission to continue and pass other routing params successfully.

**Link to Devin run**: https://app.devin.ai/sessions/9eef198322914697a9acb3bba1fa68dc
**Requested by**: joe@cal.com (@joeauyeung)

## Mandatory Tasks

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - This is a bug fix that doesn't change public APIs or user-facing behavior.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works. ✅ Existing tests pass (handleResponse.test.ts and getRoutedUrl.test.ts)

## How should this be tested?

**Environment variables needed:**
- CRM integration credentials (e.g., Salesforce, HubSpot) configured in the routing form
- `NEXT_PUBLIC_WEBAPP_URL` set

**Test scenario:**
1. Create a routing form with CRM contact owner routing enabled
2. Configure the form to route to a round-robin event type based on CRM contact owner
3. Submit the form with an email that either:
   - Doesn't exist in the CRM
   - Causes the CRM API to fail/timeout
   - Has invalid CRM credentials
4. **Expected behavior**: Form submission should succeed and redirect to the booking page with `formResponseId`, `teamMembersMatchingAttributeLogic`, and other params in the URL, even though CRM routing failed
5. **Verify**: Check logs to confirm the error was logged but not thrown

**Before this change**: Form submission would fail entirely, showing an error page
**After this change**: Form submission succeeds with null CRM values, passes other routing params

## Review Checklist

**Critical items to review:**

1. **Error handling scope**: Is it acceptable to catch all errors during CRM fetching? Should certain error types (auth, validation) be handled differently?
2. **Silent failures**: Users won't be notified when CRM routing fails (only logged). Is this acceptable, or should we surface this to users somehow?
3. **Null value behavior**: Verify that passing null for `crmContactOwnerEmail`, `crmContactOwnerRecordType`, and `crmAppSlug` correctly prevents refetching and doesn't cause issues downstream in the booking flow
4. **Logging sufficiency**: Is the error logging adequate for debugging CRM integration issues, or should we capture additional context?
5. **Test coverage**: The existing tests mock `handleResponse`, so they don't exercise this error path. Should we add a specific test for CRM fetch failures?

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings
- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)

<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Form submissions no longer fail when CRM contact owner lookup errors. We catch the error, log it, and continue with empty-string CRM values so other routing params (formResponseId, teamMembersMatchingAttributeLogic) are preserved and passed to the booking URL.

- **Bug Fixes**
  - Wrapped routerGetCrmContactOwnerEmail in a try-catch inside handleResponse. On failure or null results, log the error and set CRM fields (email, recordType, appSlug, recordId) to empty strings.
  - Forward CRM params even when empty strings and skip refetching on the booking page by checking for param presence (queries.ts and getServerSideProps.tsx).

<sup>Written for commit 17205e95cb4790eddaf8444cb87ae13441ae4983. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

devin-ai-integration[bot] committed:
fix: pass params when CRM contact owner fetch fails

When fetching CRM contact owner fails in getRoutedUrl, we now catch
the error and continue with null CRM values instead of throwing. This
ensures that other params (like formResponseId, teamMembersMatchingAttributeLogic)
are still passed to the booking URL, preventing unnecessary refetching
when the user visits the event URL.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
fix: add cal.crmFetchAttempted flag to prevent refetching

Added cal.crmFetchAttempted flag that is set when CRM contact owner
fetch is attempted in routing forms. This flag is checked on the
booking page to prevent unnecessary refetching when CRM lookup has
already been attempted (whether it succeeded or failed).

Changes:
- getRoutedUrl: Add cal.crmFetchAttempted=true when fetchCrm is true
- getServerSideProps: Check flag before refetching CRM data
- queries.ts (App Router): Check flag before refetching CRM data
- handleResponse: Catch CRM fetch errors to allow other params to pass

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Clean up

devin-ai-integration[bot] committed:
test: add tests for cal.crmFetchAttempted flag and CRM error handling

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

hariombalhara commented on apps/web/app/(booking-page-wrapper)/team/[slug]/[type]/queries.ts:
```diff
@@ -132,7 +133,10 @@ export async function getCRMData(
 
-  if (!teamMemberEmail || !crmOwnerRecordType || !crmAppSlug) {
+  const shouldSkipCrmRefetch =
+    crmFetchAttempted === "true" || (Array.isArray(crmFetchAttempted) && crmFetchAttempted[0] === "true");
```
Why do we consider crmFetchAttempted as an array ? We should assume it to be a boolean consider "true" to be only truthy value and other values to be false.

hariombalhara reviewed (changes requested):
Left a suggestion

devin-ai-integration[bot] committed:
refactor: use empty strings instead of cal.crmFetchAttempted flag

Address Hariom's feedback to avoid introducing new query params:
- Remove cal.crmFetchAttempted query parameter entirely
- Set CRM params (crmContactOwnerEmail, crmContactOwnerRecordType, crmAppSlug) to empty strings when CRM fetch fails
- Update getCRMData logic to check for param presence instead of truthiness
- Update getUrlSearchParamsToForward to include CRM params when they are not null/undefined (even if empty string)
- Update tests to expect empty strings instead of null values on CRM errors
- Fix eslint warning by changing attributeRoutingConfigParams type from any to string

This approach uses existing query params with special values (empty strings) to signal that CRM fetch was attempted but failed, avoiding the security concerns of adding a new tamperable query param.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
fix: ensure CRM params are empty strings when contact owner not found

- When fetchCrm is true and CRM fetch succeeds but returns null values, set params to empty strings instead of null
- Update getCRMData in queries.ts to treat empty strings as 'attempted and failed' and skip refetching
- Update getServerSideProps.tsx with same logic to prevent refetching when params are empty strings
- Add test case for when CRM fetch succeeds but returns null values

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into devin/1761314800-pass-params-on-crm-fetch-failure

devin-ai-integration[bot] committed:
Merge remote-tracking branch 'origin/main' into devin/1761314800-pass-params-on-crm-fetch-failure
